crates/fuel-core/src/coins_query.rs

```diff
@@ -221,6 +276,168 @@ pub async fn random_improve(
+
+    let max_dust_count = max_dust_count(max, number_of_big_coins);
+    let (dust_coins_total, selected_dust_coins) = dust_coins(
+        dust_coins_stream,
+        last_selected_big_coin,
+        max_dust_count,
+        excluded_ids,
+    )
+    .await?;
+    let retained_big_coins_iter =
+        skip_big_coins_up_to_amount(selected_big_coins, dust_coins_total);
+
+    Ok((retained_big_coins_iter.chain(selected_dust_coins)).collect())
+}
+
+async fn big_coins(
+    big_coins_stream: impl Stream<Item = StorageResult<CoinsToSpendIndexEntry>> + Unpin,
+    total: u64,
+    max: u16,
+    excluded_ids: &ExcludedCoinIds<'_>,
+) -> Result<(u64, Vec<CoinsToSpendIndexEntry>), CoinsQueryError> {
+    select_coins_until(big_coins_stream, max, excluded_ids, |_, total_so_far| {
+        total_so_far >= total
+    })
+    .await
+}
+
+async fn dust_coins(
+    dust_coins_stream: impl Stream<Item = StorageResult<CoinsToSpendIndexEntry>> + Unpin,
+    last_big_coin: &CoinsToSpendIndexEntry,
+    max_dust_count: u16,
+    excluded_ids: &ExcludedCoinIds<'_>,
+) -> Result<(u64, Vec<CoinsToSpendIndexEntry>), CoinsQueryError> {
+    select_coins_until(
+        dust_coins_stream,
+        max_dust_count,
+        excluded_ids,
+        |coin, _| coin == last_big_coin,
+    )
+    .await
+}
+
+async fn select_coins_until<F>(
+    mut coins_stream: impl Stream<Item = StorageResult<CoinsToSpendIndexEntry>> + Unpin,
+    max: u16,
+    excluded_ids: &ExcludedCoinIds<'_>,
+    predicate: F,
+) -> Result<(u64, Vec<CoinsToSpendIndexEntry>), CoinsQueryError>
+where
+    F: Fn(&CoinsToSpendIndexEntry, u64) -> bool,
+{
+    let mut coins_total_value: u64 = 0;
+    let mut coins = Vec::with_capacity(max as usize);
+    while let Some(coin) = coins_stream.next().await {
+        let coin = coin?;
+        if !is_excluded(&coin, excluded_ids)? {
+            if coins.len() >= max as usize || predicate(&coin, coins_total_value) {
+                break;
+            }
+            let amount = coin.0.amount();
+            coins_total_value = coins_total_value.saturating_add(amount);
+            coins.push(coin);
+        }
+    }
+    Ok((coins_total_value, coins))
+}
+
+fn is_excluded(
+    (key, coin_type): &CoinsToSpendIndexEntry,
+    excluded_ids: &ExcludedCoinIds,
+) -> Result<bool, CoinsQueryError> {
+    match coin_type {
+        IndexedCoinType::Coin => {
+            let utxo = key
+                .try_into()
+                .map_err(|_| CoinsQueryError::IncorrectCoinKeyInIndex)?;
+            Ok(excluded_ids.is_coin_excluded(&utxo))
+        }
+        IndexedCoinType::Message => {
+            let nonce = key
+                .try_into()
+                .map_err(|_| CoinsQueryError::IncorrectMessageKeyInIndex)?;
+            Ok(excluded_ids.is_message_excluded(&nonce))
+        }
+    }
+}
+
+fn max_dust_count(max: u16, big_coins_len: u16) -> u16 {
+    let mut rng = rand::thread_rng();
+    rng.gen_range(0..=max.saturating_sub(big_coins_len))
+}
```

Comment:
Nit: I prefer when utility functions like this one are pure. In my view, rngs should be treated as interaction points and isolated as much as possible from business logic.
```suggestion

fn max_dust_count<Rng: rand::RngCore>(max: u16, big_coins_len: u16, rng: &mut Rng) -> u16 {
    rng.gen_range(0..=max.saturating_sub(big_coins_len))
}
```

Reply:
In this case, threading the randomizer through the entire call-path was cumbersome. Also, it would need to live somewhere, probably in `FuelClient`.

As an alternative I was considering using some `OnceCell` or `OnceLock` approach, but this brought in boilerplate connected to synchronization (mutexes, thread_local!, et al.).

Since we don't really need any crypotgraphic-grade randomness here I decided to go with the simplest approach. Any suggestions are still appreciated, though.

Reply:
Fair enough :+1: